File: mQTTDevice/mQTT_conn.cs
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using Microsoft.Azure.Devices.Client;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;

namespace mQTTDevice
{
    public class deviceProps{
        public int accuracy { get; set; }
        public string battery { get; set; }
        public string client { get; set; }
        public string color { get; set; }
        public string dateTime { get; set; }
    }

    public class dataset{
        public string type { get; set; }
        public double[] coordinates { get; set; }
    }

    internal class Program
    {
        private static DeviceClient s_deviceClient;
        private static readonly TransportType s_transportType = TransportType.Mqtt;

        // The device connection string to authenticate the device with your IoT hub.
        private static string s_connectionString = "HostName=geojson-ticket-hub.azure-devices.net;DeviceId=tudelft_device001;SharedAccessKey=6dCj+Nr3TqWAuuJ303DbpEqOheoXWNKi60ixbG3Dx2Q=";

        private static async Task Main(string[] args)
        {
            Console.WriteLine("IoT Hub Quickstarts #1 - Simulated device.");

            // This sample accepts the device connection string as a parameter, if present
            ValidateConnectionString(args);

            // Connect to the IoT hub using the MQTT protocol
            s_deviceClient = DeviceClient.CreateFromConnectionString(s_connectionString, s_transportType);

            // Set up a condition to quit the sample
            Console.WriteLine("Press control-C to exit.");
            using var cts = new CancellationTokenSource();
            Console.CancelKeyPress += (sender, eventArgs) =>
            {
                eventArgs.Cancel = true;
                cts.Cancel();
                Console.WriteLine("Exiting...");
            };

            // Run the telemetry loop
            await SendDeviceToCloudMessagesAsync(cts.Token);
            await s_deviceClient.CloseAsync();

            s_deviceClient.Dispose();
            Console.WriteLine("Device simulator finished.");
        }

        private static void ValidateConnectionString(string[] args)
        {
            if (args.Any())
            {
                try
                {
                    var cs = IotHubConnectionStringBuilder.Create(args[0]);
                    s_connectionString = cs.ToString();
                }
                catch (Exception)
                {
                    Console.WriteLine($"Error: Unrecognizable parameter '{args[0]}' as connection string.");
                    Environment.Exit(1);
                }
            }
            else
            {
                try
                {
                    _ = IotHubConnectionStringBuilder.Create(s_connectionString);
                }
                catch (Exception)
                {
                    Console.WriteLine("This sample needs a device connection string to run. Program.cs can be edited to specify it, or it can be included on the command-line as the only parameter.");
                    Environment.Exit(1);
                }
            }
        }

        // Async method to send simulated telemetry
        private static async Task SendDeviceToCloudMessagesAsync(CancellationToken ct)
        {

            while (!ct.IsCancellationRequested)
            {
                // Create JSON message
                string messageBody = JsonSerializer.Serialize(
                    new
                    {
                        type = "Feature",
                        id = "54:ee:75:57:b8:59",
                        properties = new deviceProps {accuracy = 100, battery = "85%", client = "TUDelft", color = "red", dateTime = DateTime.Now.ToString("yyyy-MM-ddTHH:mm:ssZ")},
                        geometry = new dataset {type = "Point", coordinates = new[] {5.458188056945801, 51.41125257942182 }}
                    });
                using var message = new Message(Encoding.ASCII.GetBytes(messageBody))
                {
                    ContentType = "application/json",
                    ContentEncoding = "utf-8",
                };

                // Send the telemetry message
                await s_deviceClient.SendEventAsync(message);
                Console.WriteLine($"{DateTime.Now} > Sending message: {messageBody}");

                await Task.Delay(1000);
            }
        }
    }
}

File: mQTT_conn.py
# Copyright (c) Microsoft. All rights reserved.
# Licensed under the MIT license. See LICENSE file in the project root for full license information.
import asyncio
from datetime import datetime
import random
from azure.iot.device.aio import IoTHubDeviceClient
from azure.iot.device import Message

# The device connection authenticates your device to your IoT hub. The connection string for 
# a device should never be stored in code. For the sake of simplicity we're using an environment 
# variable here. If you created the environment variable with the IDE running, stop and restart 
# the IDE to pick up the environment variable.
#
# You can use the Azure CLI to find the connection string:
#     az iot hub device-identity show-connection-string --hub-name {YourIoTHubName} --device-id MyNodeDevice --output table

CONNECTION_STRING = "HostName=geojson-ticket-hub.azure-devices.net;DeviceId=tudelft_device001;SharedAccessKey=6dCj+Nr3TqWAuuJ303DbpEqOheoXWNKi60ixbG3Dx2Q="

# Define the JSON message to send to IoT Hub.
TEMPERATURE = 20.0
HUMIDITY = 60
PROPERTY_TXT = '{{"accuracy": {accuracy}, "battery": {battery}, "client": {client}, "color": {color}, "dateTime": {dateTime}}}'
GEOM_TXT = '{{"type": {typeGeom}, "coordinates": {coordinates}}}'
MSG_TXT = '{{"type": {type},"id": {id}, "properties": {properties}, "geometry": {geometry}}}' 

#   "geometry": {
#     "type": "Point",
#     "coordinates": [
#       5.458188056945801,
#       51.41125257942182
#     ]
#   }
# }

async def run_telemetry_sample(client):
    # This sample will send temperature telemetry every second
    print("IoT Hub device sending periodic messages")

    await client.connect()

    while True:
        # Build the message with simulated telemetry values.
        # temperature = TEMPERATURE + (random.random() * 15)
        # humidity = HUMIDITY + (random.random() * 20)

        type = "Feature"
        id = "54:ee:75:57:b8:59"
        accuracy = 100
        battery = "85%"
        clientel = "TUDelft"
        color = "red"
        dateTime = datetime.now().strftime("%Y-%m-%dT%H:%M:%SZ")
        typeGeom = "Point"
        coordinates = [51.43, 24.31]

        properties = PROPERTY_TXT.format(accuracy=accuracy, battery=battery, client=clientel, color=color, dateTime=dateTime)
        geometry = GEOM_TXT.format(typeGeom=typeGeom, coordinates=coordinates)

        msg_txt_formatted = MSG_TXT.format(type=type, id=id, properties=properties, geometry=geometry) #temperature=temperature, humidity=humidity, 
        message = Message(msg_txt_formatted)
        # Add a custom application property to the message.
        # An IoT hub can filter on these properties without access to the message body.
        # if temperature > 30:
        #     message.custom_properties["temperatureAlert"] = "true"
        # else:
        #     message.custom_properties["temperatureAlert"] = "false"

        # Send the message.
        print("Sending message: {}".format(message))
        await client.send_message(message)
        print("Message successfully sent")
        await asyncio.sleep(1)


def main():
    print ("IoT Hub Quickstart #1 - Simulated device")
    print ("Press Ctrl-C to exit")

    # Instantiate the client. Use the same instance of the client for the duration of
    # your application
    client = IoTHubDeviceClient.create_from_connection_string(CONNECTION_STRING)

    loop = asyncio.get_event_loop()
    try:
        # Run the sample in the event loop
        loop.run_until_complete(run_telemetry_sample(client))
    except KeyboardInterrupt:
        print("IoTHubClient sample stopped by user")
    finally:
        # Upon application exit, shut down the client
        print("Shutting down IoTHubClient")
        loop.run_until_complete(client.shutdown())
        loop.close()


if __name__ == '__main__':
    main()